pr_checkout

```diff
@@ -1,24 +1,123 @@
+#
+# * Supports copy-pasting various valid links to a pull-requests:
+# $ pr_checkout http://www.github.com/dlang/dmd/pull/1234
+# $ pr_checkout https://github.com/dlang/dmd/pull/1234/commits
+# $ pr_checkout https://github.com/dlang/dmd/pull/1234/checks
+# $ pr_checkout https://github.com/dlang/dmd/pull/1234/files
+# $ pr_checkout https://github.com/dlang/dmd/pull/9195#issuecomment-451495487
+#
+# * And even droping the obvious part of the url:
+# $ pr_checkout dlang/dmd/pull/1234
+#
+# * Use the DIR and UPSTREAM env variables to override the working directory
+#   as well as the name of the "upstream" remote:
+# $ env DIR=./dmd UPSTREAM=origin pr_checkout dlang/dmd/pull/6789
+
 set -ueo pipefail
 
 if [ -z "${1+x}" ] ; then
-    echo "Usage: ./pr_checkout <123>"
+    echo "[!] Usage: ./pr_checkout <number>|<https://github.com/<user>/<repo>/pull/<number>>"
     exit
 fi
 
-DIR="$(pwd)"
-PR="$1"
+DIR=$(test -d ${DIR:-.} && cd ${DIR:-.} || true && pwd)
 UPSTREAM="${UPSTREAM:-upstream}"
 MASTER="${MASTER:-master}"
 
+function prompt {
+    while true; do
+        read -p "$1" choice
+        case "$choice" in
+            y|Y|'') return 0;;
+            n|N) return 1;;
+            *) ;;
+        esac
+    done
+}
+
+function is_git_repo {
+    dir=$1; user=${2:-}; repo=${3:-}; upstream=${4:-};
+    if ! git -C $dir rev-parse --is-inside-work-tree >/dev/null 2>&1; then
+        return 1;
+    fi
+    if ! test -z "$user" || ! test -z "$repo" || ! test -z $upstream; then
+        repo_regex="^${upstream:-.*}[[:space:]]*\
+((https://github\.com/)|(git@github\.com:))${user:-.*}/${repo:-.*}(\.git)?"
+        git -C $dir remote -v | grep -qE "$repo_regex"
+    else
+        return 0
+    fi
+}
+
+function is_repo_clean { git -C $1 diff-index --quiet HEAD --; }
+
+function git_stash_if_needed {
+    if ! is_repo_clean $1; then
+        echo "[!] Repo '$1' contains uncommited changes."
+        if ! prompt "[?] Stash uncommitted changes [Y/n]? "; then
+            echo "[!] Unable to proceed. Good bye."; exit 1
+        fi
+        git -C $1 stash push
+    fi
+}
+
+echo "[>] Executing from '$DIR'."
+
 cd "$DIR"
+pr_num_regex='^([0-9]+)$'
+pr_url_regex='^(https?://(www\.)?github\.com/)?(.+)/(.+)/pull/([0-9]+)([/#].*)?$'
+input=$1
+if [[ $input =~ $pr_num_regex ]]; then
+    PR="${BASH_REMATCH[1]}"
+    if ! is_git_repo $DIR '' '' $UPSTREAM; then
+        echo "[!] Dir '$DIR' is not a valid git repo."
+        exit 1
+    fi
+
+    git_stash_if_needed $DIR
+    echo "[>] Attempting to fetch PR #$PR from '$UPSTREAM' ($(git remote get-url $UPSTREAM))."
+
+elif [[ $input =~ $pr_url_regex ]]; then
+    user="${BASH_REMATCH[3]}"
+    repo="${BASH_REMATCH[4]}"
+    PR="${BASH_REMATCH[5]}"
+
+    echo "[>] Attempting to fetch PR #$PR from 'github.com/$user/$repo'."
+
+    if ! is_git_repo $DIR $user $repo "$UPSTREAM"; then
+        echo "[!] Dir '$DIR' is not a valid git clone of '$user/$repo'."
+        if ! prompt "[?] Clone to '$DIR/$repo' [Y/n]? "; then
```

Comment:
Long-term wise it's probably a good idea to accept a custom string here.

Reply:
Ah, you mean asking the user where to clone? Yeah, I considered that as well, but I decided that the `DIR` env variable is enough.
If you think that it's important I can easily add it. We could also prompt the user for info on line 89, instead of bailing out.

Reply:
Yep, so any other string except for Y|y|N|n could be the repo path.